File: BaiduClub/knn-mnist/knn.py
import readFile

class Image(object) :
	__slots__ = ('num', 'length', 'data')
	def __init__(self, num, length, data) :
		self.num = num
		self.length = length
		self.data = data

class Lable(object) :
	__slots__ = ('num', 'data')
	def __init__(self, num, data) :
		self.num = num
		self.data = data

K = 20
trainImage = Image(*readFile.loadImages('train-images.idx3-ubyte'))
trainLable = Lable(*readFile.loadLables('train-labels.idx1-ubyte'))
testImage = Image(*readFile.loadImages('t10k-images.idx3-ubyte'))
testLable = Lable(*readFile.loadLables('t10k-labels.idx1-ubyte'))

def calcDist(test, train) :
	ans = 0
	for i in range(trainImage.length) :
		ans += abs((test[i] != 0) ^ (train[i] != 0))
	return ans

def calcMinplc(dist) :
	minplc = 0
	for i in range(1, len(dist)) :
		if (dist[minplc][0] > dist[i][0]) :
			minplc = i
	return minplc

def main() :
	cntFail = 0
	cntSucceed = 0
	for i in range(testImage.num) :
		dist = []
		test = testImage.data[i]
		for j in range(trainImage.num // 10) :
			train = trainImage.data[j]
			tmp = calcDist(test, train)
			dist.append([tmp, trainLable.data[j]])
#			print("successfully compared with %s" %j)
		dist.sort(key = lambda x : x[0])
#		print(dist)
		cnt = [0] * 15
		maxplc = 1
		for j in range(K) :
			cnt[dist[j][1]] += 1
			if (cnt[dist[j][1]] > cnt[maxplc]) :
				maxplc = dist[j][1]
		if (maxplc == testLable.data[i]) :
			cntSucceed += 1
		else :
			cntFail += 1
		print("Succeed %d," %cntSucceed, "Fail %d" %cntFail)
	print(float(cntSucceed / testImage.num))

main()

File: BaiduClub/knn-mnist/knn.cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 6e4 + 10;
const int M = 1e4 + 10;
const int L = 800;
const int K = 20;

int n, m, len;
char ate[M], atr[N];
char ste[M][L], str[N][L];

void init() {
	freopen("testImage.in", "r", stdin);
	scanf("%d", &m);
	for (int i = 1; i <= m; ++i)
		scanf("%s", ste[i]);
	freopen("testLable.in", "r", stdin);
	scanf("%d %s", &m, ate + 1);

	freopen("trainImage.in", "r", stdin);
	scanf("%d", &n);
	for (int i = 1; i <= n; ++i)
		scanf("%s", str[i]);
	freopen("trainLable.in", "r", stdin);
	scanf("%d %s", &n, atr + 1);
	len = strlen(ste[1]);
}

struct nodeHeap {
	int dist, ans;
	nodeHeap() {}
	nodeHeap(int a, int b) : dist(a), ans(b) {}
	inline bool operator < (const nodeHeap &a) const {
		return dist < a.dist;
	}
};
priority_queue <nodeHeap> H;

#define id(a) (a - '0')
inline int calcDist(int a, int b) {
	int ans = 0;
	for (int i = 1; i <= len; ++i)
		ans += abs(ste[a][i] - str[b][i]);
	return ans;
}
bool judge(int a) {
	while (!H.empty()) H.pop();
	for (int i = 1; i <= n; ++i) {
		nodeHeap tmp = nodeHeap(calcDist(a, i), id(atr[i]));
		if (H.size() < K) H.push(tmp);
		else if (tmp < H.top()) {
			H.pop(), H.push(tmp);
		}
	}
	int cnt[10], maxn = 0;
	memset(cnt, 0, sizeof(cnt));
	while (!H.empty()) {
		int ans = H.top().ans; H.pop();
		if (++cnt[ans], cnt[ans] > cnt[maxn])
			maxn = ans;
	}
	return maxn == id(ate[a]);
}

int main() {
	init();
	int cntS = 0, cntF = 0;
	for (int i = 1; i <= m; ++i) {
		if (judge(i))
			++cntS;
		else ++cntF;
		printf("Succeed %d, Fail %d\n", cntS, cntF);
	}
	printf("%lf\n", (double)cntS / (cntS + cntF));
	return 0;
}

